### PR TITLE
Avoid adding a closed client to the pool

### PR DESCRIPTION
**Pull request by flyrain**
cc @RussellSpitzer @aokolnychyi @szehon-ho @amogh-jahagirdar @nastra

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/ClientPoolImpl.java [hunk 1]
```diff
@@ -70,37 +71,29 @@ public <R> R run(Action<R, C, E> action, boolean retry) throws E, InterruptedExc
     try {
       return action.run(client);
     } catch (Exception exc) {
-      if (!retry || !isConnectionException(exc)) {
-        throw exc;
+      if (retry && isConnectionException(exc)) {
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/ClientPoolImpl.java:74]: Didn't find a good to wrap the logic in a method, but open for suggestions

File: core/src/main/java/org/apache/iceberg/ClientPoolImpl.java [hunk 2]
```diff
@@ -70,37 +71,29 @@ public <R> R run(Action<R, C, E> action, boolean retry) throws E, InterruptedExc
...
+            }
+          }
+        }
       }
 
-      return retryAction(action, exc, client);
-
+      throw exc;
     } finally {
       release(client);
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/ClientPoolImpl.java:93]: This adds the client back to the pool. We should avoid adding a closed client back to the pool.

**Review by amogh-jahagirdar — APPROVED** [review]
Thanks for fixing this @flyrain ! The change looks good to me.

**Comment by flyrain** [comment]
Thanks @amogh-jahagirdar for the review!

**Pull request merged**